cpp/src/arrow/columnar_format.cc

```diff
@@ -0,0 +1,1034 @@
+    *node_out = std::move(node);
+    return Status::OK();
+  }
+
+  Status Handle(const Array& array, int64_t i, int16_t rep_level) override {
+    return Handle(checked_cast<const ListArray&>(array), i, rep_level);
+  }
+
+  Status Handle(const ListArray& array, int64_t i, int16_t rep_level) {
+    if (array.IsNull(i)) {
+      RETURN_NOT_OK(AppendLevelsRecursive(rep_level, int16_t(def_level_ - 2)));
+    } else if (array.value_length(i) == 0) {
+      RETURN_NOT_OK(AppendLevelsRecursive(rep_level, int16_t(def_level_ - 1)));
+    } else {
+      RETURN_NOT_OK(rep_levels_builder_->Append(rep_level));
+      RETURN_NOT_OK(def_levels_builder_->Append(def_level_));
+      int32_t offset = array.value_offset(i);
+      RETURN_NOT_OK(child(0)->Handle(*array.values(), offset, rep_level));
+      for (int64_t j = 1; j < array.value_length(i); j++) {
+        RETURN_NOT_OK(child(0)->Handle(*array.values(), offset + j, rep_level_));
+      }
+    }
+    return Status::OK();
+  }
+
+  // only levels
+  Status Finish(ColumnMap* colmap) override {
+    std::shared_ptr<Array> rep_levels_tmp;
+    std::shared_ptr<Array> def_levels_tmp;
+    RETURN_NOT_OK(rep_levels_builder_->Finish(&rep_levels_tmp));
+    RETURN_NOT_OK(def_levels_builder_->Finish(&def_levels_tmp));
+    colmap->Put(field_, std::static_pointer_cast<Int16Array>(rep_levels_tmp),
+                        std::static_pointer_cast<Int16Array>(def_levels_tmp),
+                        nullptr);
+    return Status::OK();
+  }
+};
+
+Status ShredNode::CreateTree(const std::shared_ptr<Field>& field,
+                             int16_t rep_level, int16_t def_level,
+                             MemoryPool* pool,
+                             std::unique_ptr<ShredNode>* node) {
+#define CASE(data_type) \
+    case data_type::type_id: \
+      return PrimitiveShredNode<data_type>::Create(field, rep_level, def_level, pool, node);
+  switch (field->type()->id()) {
+    case Type::STRUCT:
+      return StructShredNode::Create(field, rep_level, def_level, pool, node);
+    case Type::LIST:
+      return ListShredNode::Create(field, rep_level, def_level, pool, node);
+    PRIMITIVE_CASE_LIST
+    default:
+      return Status::Invalid("Unsupported type: ", field->type()->name());
+  }
+#undef CASE
+}
+
+} // anonymous namespace
+
+class Shredder::Impl {
+ public:
+  Impl(std::unique_ptr<ShredNode> root) : root_(std::move(root)) {}
+
+  std::unique_ptr<ShredNode> root_;
+};
+
+Status Shredder::Create(std::shared_ptr<Field> schema,
+                        MemoryPool* pool,
+                        std::shared_ptr<Shredder>* shredder) {
+  std::unique_ptr<ShredNode> root;
+  RETURN_NOT_OK(ShredNode::CreateTree(schema, 0, 0, pool, &root));
+
+  std::unique_ptr<Impl> impl(new Impl(std::move(root)));
+
+  shredder->reset(new Shredder(std::move(impl)));
+  return Status::OK();
+}
+
+Shredder::Shredder(std::unique_ptr<Impl> impl) : impl_(std::move(impl)) {
+}
+
+Shredder::~Shredder() {
```

Comment:
just use default at declaration point?

Reply:
It won't compile. `sizeof(Impl)` is not known at declaration point.